File: src/main/source/com/toocol/lc/cpp/join_search_set.cpp
#include <vector>
#include <numeric>

using namespace std;

class JoinSearchSet
{
private:
    vector<int> parent;
    vector<int> rank;

public:
    int maxSize = 1;
    JoinSearchSet(int capacity)
        : parent(capacity, -1),
          rank(capacity, 1)
    {
    }

    int findRoot(int x)
    {
        while (parent[x] != -1)
            x = parent[x];
        return x;
    }

    void joinit(int x, int y)
    {
        int rx = findRoot(x), ry = findRoot(y);
        if (rx == ry)
            return;
        if (rank[rx] > rank[ry])
        {
            parent[ry] = rx;
            rank[rx] += rank[ry];
            maxSize = max(maxSize, rank[rx]);
        }
        else
        {
            parent[rx] = ry;
            rank[ry] += rank[rx];
            maxSize = max(maxSize, rank[ry]);
        }
    }
};

typedef JoinSearchSet JSS;

File: src/main/source/com/toocol/lc/python/join_search_set.py
# -*- encoding: utf-8 -*-
# @date    : 2022/07/17 12:47:08
# @author  : Joe Zane
# @email   : joezane.cn@gmail.com
# @version : v1.0


class JoinSearchSet:
    def __init__(self):
        self.parent, self.rank = [], []
        self.maxSize = 0

    def initialise(self, capacity: int):
        self.parent = [-1] * capacity
        self.rank = [1] * capacity

    def findRoot(self, x: int) -> int:
        while self.parent[x] != -1:
            x = self.parent[x]
        return x

    def joint(self, x: int, y: int):
        rx, ry = self.findRoot(x), self.findRoot(y)
        if rx == ry:
            return
        if self.rank[rx] > self.rank[ry]:
            self.parent[ry] = rx
            self.rank[rx] += self.rank[ry]
            self.maxSize = max(self.maxSize, self.rank[rx])
        else:
            self.parent[rx] = ry
            self.rank[ry] += self.rank[rx]
            self.maxSize = max(self.maxSize, self.rank[ry])
